drivers/clock_control/nrf_power_clock.c

```diff
@@ -226,68 +188,151 @@ static inline void anomaly_132_workaround(void)
 #endif
 }
 
-static int clock_async_start(struct device *dev,
-			     clock_control_subsys_t subsys,
-			     struct clock_control_async_data *data)
+static int async_start(struct device *dev, clock_control_subsys_t subsys,
+			struct clock_control_async_data *data, uint32_t ctx)
 {
 	enum clock_control_nrf_type type = (enum clock_control_nrf_type)subsys;
-	const struct nrf_clock_control_sub_config *config;
-	struct nrf_clock_control_sub_data *clk_data;
-	int key;
-	uint8_t ref;
+	struct nrf_clock_control_sub_data *subdata = get_sub_data(dev, type);
+	int err;
 
-	__ASSERT_NO_MSG(type < CLOCK_CONTROL_NRF_TYPE_COUNT);
-	config = get_sub_config(dev, type);
-	clk_data = get_sub_data(dev, type);
-
-	__ASSERT_NO_MSG((data == NULL) ||
-			((data != NULL) && (data->cb != NULL)));
-
-	/* if node is in the list it means that it is scheduled for
-	 * the second time.
-	 */
-	if ((data != NULL)
-	    && is_in_list(&clk_data->list, &data->node)) {
-		return -EBUSY;
+	err = set_state(&subdata->flags, CLOCK_CONTROL_STATUS_OFF,
+			ctx | CLOCK_CONTROL_STATUS_STARTING);
+	if (err < 0) {
+		return err;
```

Comment:
AFAICT this will return `-EFAULT` if an already started (or starting) clock is started again by a call to `clock_control_on()` or `clock_control_async_on()` (or maybe even `-EPERM` if `subdata->flags` for some reason got changed between the call to `atomic_cas()` and later comparison of the flags, what seems to be not impossible). Previous implementation returned 0 in such case. Is that an intentional change in behavior?

Reply:
i've rewrote setting state. `-EBUSY` is returned if on is requested (blocking or async) and clock is not in off state. We need clearance in the API. Since we are removing reference counting from clock_control API then second on request should always error (`-EBUSY`).

The problem is the order. I would prefer that this PR goes in which will violate API in `clock_control_async_on` by returning `-EBUSY` when clock is already running (API states that it should return 0) but:
- it is used only by nordic.
- nordic will no longer use it directly after switch to onoff service.

Reply:
> We need clearance in the API. Since we are removing reference counting from clock_control API then second on request should always error (`-EBUSY`).

I'm not convinced we should do it. What about other clock_control drivers? They don't seem to check if the clock was already requested. Should they all be updated then?
But why actually the second request should be considered an error?

Reply:
> But why actually the second request should be considered an error?

Since `calling clock_control_off()` shuts down the clock (without reference counting) then imo it should be only one originator of clock being on. Wouldn't it be error prone? If there are multiple sources which turns clock on and off then it calls for onoff and it might go unnoticed if `clock_control_on` returns success when enabling already enabled clock. imo, on and off should return  `-EALREADY` (better than `-EBUSY`) when clock is in unexpected state. 

> What about other clock_control drivers? 

Note that api clock control is still poor defined with no description of errors in `clock_control_on` and `clock_control_off`. Another thing is that typically (always?) clock control is used by vendor dependent layer so it is not to give hardware abstraction but rather common look and feel (and test suite).

Reply:
Agreed. This will indeed help identify cases where switching to using the on-off service is needed. And yes, `-EALREADY` seems more suitable for both `clock_control_on()` and `clock_control_off()`. But I don't think it will help if we mention that in the API description. Let's stay with this opaque "0 on success, negative errno on failure". This allows more flexible approach to be used by different vendors, what is actually good for such vendor specific thing like clock handling.

Reply:
If you want opaque and flexible use non-negative for success so vendors can convey additional information (like whether the request was unnecessary) in the non-error return value.